### PR TITLE
Add @__DIR__ and document use in pkgs. Fix #12120

### LINKED ISSUE #12120: RFC: fix package usages of Pkg.dir()
Many packages use `Pkg.dir()` to find load static resources that aren't Julia code. However, `Pkg.dir()` always points to `~/.julia/v0.x`, so this breaks for packages which are stored in `julia/usr/share/julia/site/v0.x`, for example. It looks like we need a better way to handle this if there isn't one already.

Are packages that use `Pkg.dir()` for this purpose in the wrong, or should `Pkg.dir()` be smarter? Or should we just only support a single package directory?

### PR DESCRIPTION
**Pull request by helgee**
x-ref: #12120

### CODE REVIEW

**Comment by helgee** [comment]
Added tests and docs.

File: doc/manual/packages.rst [hunk 1]
```diff
@@ -771,6 +771,16 @@ testing service, and an ``appveyor.yml`` file for using `AppVeyor <https://www.a
...
 
+Loading Static Non-Julia Files
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If your package code needs to load static files which are not Julia code, e.g. an external library or data files, and are located
+within the package directory, use the ``@__DIR__`` macro to determine the directory of the current source file. For example if
+package ``FooBar`` needs to load ``foo.csv`` from the ``data`` directory, use the following code::
+
+    datapath = joinpath(@__DIR__, "..", "data")
+    foo = readcsv(joinpath(datapath, "foo.csv")
```
Comment (simonster) [inline doc/manual/packages.rst]: Missing `)`

File: base/docs/helpdb/Base.jl [hunk 2]
```diff
@@ -2374,6 +2374,15 @@ macro. Returns `nothing` if run from a REPL or an empty string if evaluated by
 :@__FILE__
 
 """
+    @__DIR__ -> AbstractString
+
+`@__DIR__` expands to a string with the directory part of the path of the file containing
+the macro. Returns `nothing` if run from a REPL or an empty string if evaluated by
+`julia -e <expr>`.
+"""
+:@__DIR__
```
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: could you put this inline? we're trying to move docs out of helpdb

File: base/docs/helpdb/Base.jl [hunk 3]
```diff
@@ -2374,6 +2374,15 @@ macro. Returns `nothing` if run from a REPL or an empty string if evaluated by
 :@__FILE__
 
 """
+    @__DIR__ -> AbstractString
+
+`@__DIR__` expands to a string with the directory part of the path of the file containing
```
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: may as well specify that it's an absolute path

**Comment by helgee** [comment]
Clumsy merge. But this will get squashed anyways, am I right?

**Comment by tkelman** [comment]
it should if you don't want to do it yourself (best to make pr's from branches other than master of your fork in the future)

**Comment by helgee** [comment]
@tkelman Forgetting to branch belongs to the things that happen in those quick-fix PRs past one's bedtime. I promise to be fully awake next time. 😉

**Comment by tkelman** [comment]
oh and the squash merge should remember to remove the [ci skip]s from the intermediate commits since you do want ci to run on the merge.

**Comment by helgee** [comment]
Squashed and CI passed.

**Comment by tkelman** [comment]
Nice work @helgee. Good candidate to add to Compat so packages can start using it.

**Comment by helgee** [comment]
@tkelman https://github.com/JuliaLang/Compat.jl/pull/281

**Pull request merged**